File: src/SpiralMatrix.java
/**
 * 
 * SprilMatrix.java
 * Dec 5, 2012
 */

/**
 * @author huangsp
 *
 */
public class SpiralMatrix {
	
	public static void print(int[][] num){
		int[][] direction = {{0,1},{1,0},{0,-1},{-1,0}};
		boolean[][] visited = new boolean[num.length][num[0].length];
		int ROW = num.length;
		int COL = num[0].length;
		int row = 0, col = 0, i = 0, j = 0, m = 0;
		
		while (!visited[i][j]){
			row = i + direction[m][0];
			col = j + direction[m][1];
			if (row < ROW && row >= 0 && col < COL && col >= 0 && !visited[row][col]){
				System.out.printf("%d ",num[i][j]);
				visited[i][j] = true;
				i = row;
				j = col;
			}else{
				m++;
				if (m > 3)	m = 0;
				row = i + direction[m][0];
				col = j + direction[m][1];
				if (visited[row][col]){
					visited[i][j] = true;
					System.out.println(num[i][j]);	
				}
			}
		}
	}
	public static int[][] set(int N, int M){
		int[][] num = new int[N][M];
		int[][] direction = {{0,1},{1,0},{0,-1},{-1,0}};
		int current = 1;
		int row = 0;
		int col = 0;
		int i = 0, j = 0, m = 0;
		while (current <= N * M){
			row = i + direction[m][0];
			col = j + direction[m][1];
			if (row < N && row >= 0 && col < M && col >= 0 && num[row][col] == 0){
				num[i][j] = current++;
				i = row;
				j = col;
			}else{
				m++;
				if (m > 3) m = 0;
				row = i + direction[m][0];
				col = j + direction[m][1];
				if (num[row][col] != 0){
					num[i][j] = current++;
				}
			}
		}
		return num;
	}
	public static int sqrt(int x){
		if ( x < 2)		return x;
		else{
			double root = x;
			// Newton's recursive method
			for (int i = 0; i < 20; i++){
				root = (root + x / root) /2;
			}
			return (int)(root);
		}
	}
	public static void main(String[] args){
		@SuppressWarnings("unused")
		int matrix[][] = {{1,2,3,4,5},
				{16,17,18,19,6},
				{15,24,25,20,7},
				{14,23,22,21,8},
				{13,12,11,10,9}};
		print (set(5,4));
		//System.out.println(sqrt(16));
	}
}

File: python/spiral_matrix.py
#! /usr/bin/python

'''
Given an integer n, generate a square matrix filled with elements from 1 to n2 in spiral order.

For example,
Given n = 3,
You should return the following matrix:
[
 [ 1, 2, 3 ],
 [ 8, 9, 4 ],
 [ 7, 6, 5 ]
]

Follow up:
Given the following matrix:
[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
You should return [1,2,3,6,9,8,7,4,5].
'''

class Solution:
    # @return a list of lists of integer
    def generateMatrix(self, n):
        num = 1
        matrix = [ [0] * n for index in range(n) ]
        direction = [ (0,1), (1,0), (0,-1), (-1,0) ]
        size = n * n
        row = 0
        col = -1
        direct_index = 0
        while num <= size:
            next_row = row + direction[direct_index][0]
            next_col = col + direction[direct_index][1]
            if next_row < n and next_col < n and matrix[next_row][next_col] == 0:
                row = next_row
                col = next_col
                matrix[row][col] = num
                num += 1
            else:
                direct_index += 1
                direct_index %= 4
        return matrix

    # @param matrix, a list of lists of integers
    # @return a list of integers
    def spiralOrder(self, matrix):
        output = list()
        if not matrix:
            return output
        ROW = len(matrix)
        COL = len(matrix[0])
        SIZE = ROW * COL
        direct_index = 0
        direction = [ (0,1), (1,0), (0,-1), (-1,0) ]
        curr_row = 0
        curr_col = -1
        index = 0
        VISITED = -1e10
        while index < SIZE:
            next_row = curr_row + direction[direct_index][0]
            next_col = curr_col + direction[direct_index][1]
            if next_row < ROW and next_col < COL and matrix[next_row][next_col] != VISITED:
                curr_row = next_row
                curr_col = next_col
                output.append(matrix[curr_row][curr_col])
                matrix[curr_row][curr_col] = VISITED
                index += 1
            else:
                direct_index += 1
                direct_index %= 4
        return output
if __name__ == '__main__':
    solution = Solution()
    matrix = solution.generateMatrix(5)
    print solution.spiralOrder(matrix)
